### PR TITLE
Entity form : Establish methodology to allow default values to be passed in on the url

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Establish methodology to allow default values to be passed in on the url where the new EntityForm trait is used & by extension later other forms that use the fieldsArray format

Before
----------------------------------------
Cannot accept values from the url 

After
----------------------------------------
Very limited ability to accept values from the url  civicrm/admin/member/membershipStatus?action=update&reset=1&label=bob&is_admin=1
pre-fills **label** & **is_admin**

Technical Details
----------------------------------------
This is an extension of the standard developed in https://lab.civicrm.org/dev/core/issues/115 with the intent of determining a safe, easy to implement way, to allow variables to be passed via the url. It relies on some cleanup & I hope https://github.com/civicrm/civicrm-core/pull/12184 will be merged allowing me to rebase this to the essentials.

Basically the idea is that as long as that values coming in from the url are safe provided they are only letters, numbers & _. We should accept those 'easily' (but leave anything that can't work within that out of spec). Otherwise we risk people adding url-acceptable fields in more ad hoc ways.

Comments
----------------------------------------

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by JoeMurray** [comment]
Only concern is whether this might interact with current consumption of url key-values somewhere. About 5-8 years ago I put some support in for passing params to, hmm, membership report, so the filter values for membership type and period could be set from url in order to allow clicking from values displayed on an earlier version of Membership Dashboard.

**Comment by colemanw** [comment]
@eileenmcnaughton this looks like it may duplicate work by @seanmadsen to do alphanumeric validation.

**Comment by eileenmcnaughton** [comment]
@colemanw it is a function @seamuslee001 wrote for another purpose - but it is not yet merged - hence included here

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
@colemanw wanna discuss this at some point? I wanna resolve it as a step towards my concerns in https://github.com/civicrm/civicrm-core/pull/11055#issuecomment-395931798

**Comment by MegaphoneJon** [comment]
I've received funding for an extension to create "template" functionality for contacts/contributions, similar to event templates.  Rather than store a "template" record, I had instead intended to store the values in JSON in `civicrm_setting` (or possibly a new entity), with an EntityRef to select them on the "New Contact"/"New Contribution" screens.

Conceivably, we could replace the stored entities with URL params - but we'd lose the AJAX-y aspect of it that makes for nice UX.  If someone has a clever idea for how we can embed URL params and not do another page load, I'm happy to redirect funding toward that. Ping me on Mattermost if so!

**Comment by eileenmcnaughton** [comment]
@MegaphoneJon you saw https://wiki.civicrm.org/confluence/display/CRM/Configurable+Contact+Summary+Screen didn't you? I feel like you discussed it on chat

**Comment by MegaphoneJon** [comment]
@eileenmcnaughton I did see the Configurable Contact Summary Screen - and while that's very interesting to me, in this context the "template" I'm referring to is the ability to pre-populate values on, say, the "New Individual" screen.  Passing the values via URL is an intriguing option.

**Comment by eileenmcnaughton** [comment]
@colemanw part of why I wanted to resolve this is so we have a template for other forms - we now have 2 PRs open relating to passing values by url from search (e.g #12455) . This is an area where bad things can happen so I want to try to get a metadata-based, explicit approach to values from urls.

File: CRM/Core/Form/EntityFormTrait.php [hunk 1]
```diff
@@ -131,6 +131,47 @@ protected function addFormButtons() {
...
+   * @param string $fieldName
+   *
+   * @return string|int|bool
+   */
+  protected function getValidationTypeForField($fieldName) {
+    switch ($this->metadata[$fieldName]['type']) {
+      case CRM_Utils_Type::T_BOOLEAN:
+        return 'Boolean';
+
+      default:
```
Comment (eileenmcnaughton) [inline CRM/Core/Form/EntityFormTrait.php]: I admit this part is not very fleshed out.... but Alphanumeric is 'safe'

**Comment by eileenmcnaughton** [comment]
@colemanw are you able to look at this one - would really like your input. There are 2 other PRs open around handling url params is search & I'd like to think we could extend this idea to those (I'm fairly committed to insisting on a metadata approach to accepting new params-by-url in search for security if nothing else)

File: CRM/Core/Form/EntityFormTrait.php (hunk 1 above)
Comment (colemanw) [reply CRM/Core/Form/EntityFormTrait.php]: Yea I feel like Alphanumeric is too restrictive because it doesn't allow e.g. spaces. (So "Bob" would be accepted but "Bob Jones" would be rejected for no reason). But this is a good start!

**Comment by colemanw** [comment]
I've tested this out and can confirm that it works.

**Comment by colemanw** [comment]
* ([`r-jira`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-jira)) __Pass__
* ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) __Pass__
* ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) __Pass__
* ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) __Pass__
* ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) __Pass__
* ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) __Pass__
* ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) __Pass__
* ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) __Pass__
